src/Api/PubnubApi/EndPoint/Files/SendFileOperation.cs

```diff
@@ -2,541 +2,574 @@
+				PNStatus status = new StatusBuilder(config, jsonLibrary).CreateStatusResponse(PNOperationType.PNFileUploadOperation, category, requestState, statusCode, new PNException(transportResponse.Error.Message, transportResponse.Error));
+				JsonAndStatusTuple = new Tuple<string, PNStatus>(string.Empty, status);
+			}
+			returnValue.Status = JsonAndStatusTuple.Item2;
+			string json = JsonAndStatusTuple.Item1;
+			if (!string.IsNullOrEmpty(json)) {
+				LoggingMethod.WriteToLog(pubnubLog, string.Format(CultureInfo.InvariantCulture, "DateTime {0} GenerateFileUploadUrl -> file upload OK.", DateTime.Now.ToString(CultureInfo.InvariantCulture)), config.LogVerbosity);                //do internal publish after successful file upload
+
+				Dictionary<string, object> publishPayload = new Dictionary<string, object>();
+				if (this.publishFileMessageContent != null && !string.IsNullOrEmpty(this.publishFileMessageContent.ToString())) {
+					publishPayload.Add("message", this.publishFileMessageContent);
+				}
+				currentFileId = generateFileUploadUrlResult.FileId;
+				sendFileName = generateFileUploadUrlResult.FileName;
+				publishPayload.Add("file", new Dictionary<string, string> {
+						{ "id", generateFileUploadUrlResult.FileId },
+						{ "name", generateFileUploadUrlResult.FileName } });
+
+				int publishFileRetryLimit = config.FileMessagePublishRetryLimit;
+				int currentFileRetryCount = 0;
+				bool publishFailed;
+				do {
+					currentFileRetryCount += 1;
+					PNResult<PNPublishFileMessageResult> publishFileMessageResponse = await PublishFileMessage(publishPayload, queryParam).ConfigureAwait(false);
+					PNPublishFileMessageResult publishFileMessage = publishFileMessageResponse.Result;
+					PNStatus publishFileMessageStatus = publishFileMessageResponse.Status;
+					if (publishFileMessageStatus != null && !publishFileMessageStatus.Error && publishFileMessage != null) {
+						publishFailed = false;
+						PNFileUploadResult result = new PNFileUploadResult();
+						result.Timetoken = publishFileMessage.Timetoken;
+						result.FileId = generateFileUploadUrlResult.FileId;
+						result.FileName = generateFileUploadUrlResult.FileName;
+						returnValue.Result = result;
+						returnValue.Status.Error = false;
+						LoggingMethod.WriteToLog(pubnubLog, string.Format(CultureInfo.InvariantCulture, "DateTime {0} GenerateFileUploadUrl -> file upload -> PublishFileMessage -> Success.", DateTime.Now.ToString(CultureInfo.InvariantCulture)), config.LogVerbosity);                //do internal publish after successful file upload
+					} else {
+						publishFailed = true;
+						returnValue.Status = publishFileMessageStatus;
+						LoggingMethod.WriteToLog(pubnubLog, string.Format(CultureInfo.InvariantCulture, "DateTime {0} PublishFileMessage Failed. currentFileRetryCount={1}", DateTime.Now.ToString(CultureInfo.InvariantCulture), currentFileRetryCount), config.LogVerbosity);
+						Task.Delay(1000).Wait();
```

Comment:
Why not await Task.Delay() here?

Reply:
Yes, Missing refactoring thing
Updated!